airbyte-workers/src/main/java/io/airbyte/workers/internal/EmptyAirbyteSource.java

```diff
@@ -61,4 +120,109 @@ public void cancel() throws Exception {
+                .withNamespace(configuredAirbyteStream.getStream().getNamespace()))
+        .collect(Collectors.toSet());
+    final Set<StreamDescriptor> configStreamDescriptors = new HashSet<>(streamDescriptors);
+
+    return catalogStreamDescriptors.equals(configStreamDescriptors);
+  }
+
+  private AirbyteMessage getNullStreamStateMessage(final StreamDescriptor configStreamDescriptor) {
+    return new AirbyteMessage()
+        .withType(Type.STATE)
+        .withState(
+            new AirbyteStateMessage()
+                .withStateType(AirbyteStateType.STREAM)
+                .withStream(
+                    new AirbyteStreamState()
+                        .withStreamDescriptor(new io.airbyte.protocol.models.StreamDescriptor()
+                            .withName(configStreamDescriptor.getName())
+                            .withNamespace(configStreamDescriptor.getNamespace()))
+                        .withStreamState(null)));
+  }
+
+  private AirbyteMessage getNullGlobalMessage(final Queue<StreamDescriptor> configStreamDescriptors, final AirbyteStateMessage currentState) {
+    final AirbyteGlobalState globalState = new AirbyteGlobalState();
+    globalState.setStreamStates(new ArrayList<>());
+
+    currentState.getGlobal().getStreamStates().forEach(exitingState -> globalState.getStreamStates()
+        .add(
+            new AirbyteStreamState()
+                .withStreamDescriptor(exitingState.getStreamDescriptor())
+                .withStreamState(
+                    configStreamDescriptors.contains(new StreamDescriptor()
+                        .withName(exitingState.getStreamDescriptor().getName())
+                        .withNamespace(exitingState.getStreamDescriptor().getNamespace())) ? null : exitingState.getStreamState())));
+
+    // If all the stream in the current state have been reset, we consider to shared as reset as well
+    if (currentState.getGlobal().getStreamStates().size() == globalState.getStreamStates().stream()
+        .filter(streamState -> streamState.getStreamState() == null).count()) {
+      log.info("All the streams of a global state have been reset, the shared state will be erased as well");
+      globalState.setSharedState(null);
+    } else {
+      log.info("This is a partial reset, the shared state will be preserve");
+      globalState.setSharedState(currentState.getGlobal().getSharedState());
+    }
+
+    // Add state being reset that are not in the current state. This is made to follow the contract of
+    // the global state always containing the entire
+    // state
+    configStreamDescriptors.forEach(configStreamDescriptor -> {
+      if (!currentState.getGlobal().getStreamStates().stream().map(streamState -> streamState.getStreamDescriptor()).toList()
+          .contains(new io.airbyte.protocol.models.StreamDescriptor()
+              .withName(configStreamDescriptor.getName())
+              .withNamespace(configStreamDescriptor.getNamespace()))) {
```

Comment:
I think this would be a bit more readable if you saved this whole block
```
new io.airbyte.protocol.models.StreamDescriptor()
              .withName(configStreamDescriptor.getName())
              .withNamespace(configStreamDescriptor.getNamespace())
```
to a variable inside of this lambda. That way you also don't need to repeat that twice